Add Muse writer

ilabdsf opened the pull request:
Closes #3405

jgm commented on pandoc.cabal:
```diff
@@ -180,6 +181,7 @@ Extra-Source-Files:
                  test/writer.markdown
                  test/writer.plain
                  test/writer.mediawiki
+                 test/writer.muse
```
I think you need to add `test/writer.muse` to the repository -- I don't see it here in this PR.

ilabdsf replied:
I don't have it actually. I added it when I began development, but removed it then as I thought it was deprecated as it was referred from `test/Tests/Old.hs`. Instead I made a comprehensive unit test suite.

Should I add it or are we moving towards unit tests?

jgm commented on README.md:
```diff
@@ -26,7 +26,7 @@ write plain text, [Markdown], [CommonMark], [PHP Markdown Extra],
 markup], [DokuWiki markup], [ZimWiki markup], [Haddock markup],
 [EPUB] \(v2 or v3\), [FictionBook2], [Textile], [groff man] pages,
-[Emacs Org mode], [AsciiDoc], [InDesign ICML], [TEI Simple], and [Slidy],
+[Emacs Org mode], [AsciiDoc], [InDesign ICML], [TEI Simple], [Muse] and [Slidy],
```
Let's put a comma after `[Muse]` here; it makes it clearer that it's not one of the slide show formats.

jgm commented on pandoc.cabal:
```diff
@@ -24,7 +24,7 @@ Description:     Pandoc is a Haskell library for converting from one markup
                  groff man pages, plain text, Emacs Org-Mode, AsciiDoc,
                  Haddock markup, EPUB (v2 and v3), FictionBook2, InDesign ICML,
-                 and several kinds of HTML/javascript slide shows (S5, Slidy,
+                 Muse and several kinds of HTML/javascript slide shows (S5, Slidy,
```
Comma after Muse please.

jgm commented on deb/control.in:
```diff
@@ -16,5 +16,5 @@ Description: general markup converter
  DokuWiki, Textile, groff man pages, plain text, Emacs Org-Mode,
  AsciiDoc, Haddock markup, EPUB (v2 and v3), FictionBook2,
- InDesign ICML, and several kinds of HTML/javascript
+ InDesign ICML, Muse and several kinds of HTML/javascript
```
Comma after Muse

jgm commented on src/Text/Pandoc/Writers/Muse.hs:
```diff
@@ -0,0 +1,329 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-
+Copyright (C) 2010-2017 John MacFarlane <jgm@berkeley.edu>
```
You should put your own name and 2017 here.

jgm commented:
This basically looks great.  I've made a couple of minor comments.
The biggest problem is the missing test/writer.muse.

ilabdsf commented:
@jgm I have fixed what you proposed.

As for `test/writer.muse`, I have just removed it as I suggested in the comment above.

Also, just in case you have not noticed, I have not added Muse into the list of formats in `man/pandoc.1` because I think it is regenerated anyway when release is made.

jgm commented:
I think we should add it.  Unit tests are still good, but
we have both for a number of the other writers.  Since this
test converts a document containing most document features,
it gives good coverage.

You can generate it easily.  Just add an empty
test/writer.muse, and run the test suite.  It will fail
and tell you the exact command you can use to regenerate
test/writer.muse.  Generate it and check it very carefully
for correctness.  If everything looks good, commit it.

jgm commented:
+++ Alexander Krotov [Mar 06 17 11:39 ]:
>   Also, just in case you have not noticed, I have not added Muse into the
>   list of formats in man/pandoc.1 because I think it is regenerated
>   anyway when release is made.

Correct.

ilabdsf commented:
Found some bugs to fix:
- [ ] Definition lists are not intended properly.
- [ ] Accidental lists are not escaped, should probably detect and mark `1.` and such as `<verbose>`.
- [ ] Loose lists become tight lists.

jgm commented:
Let me know when you're ready to merge. I wasn't sure if you were planning to do something about the accidental lists issue.

ilabdsf commented:
@jgm
Ok, lets merge it as is. These issues are indeed minor, especially because Muse requires whitespace before list markers.

I will not change this branch anymore and push all new changes into another branch, like `muse-next`.
